web-common/src/features/dashboards/stores/persistent-dashboard-state.ts

```diff
@@ -0,0 +1,79 @@
+import type {
+  SortDirection,
+  SortType,
+} from "@rilldata/web-common/features/dashboards/proto-state/derived-types";
+import { localStorageStore } from "@rilldata/web-common/lib/store-utils";
+import { get, type Readable, type Updater } from "svelte/store";
+
+/**
+ * Partial state of the dashboard that is stored in local storage.
+ */
+export type PersistentDashboardState = {
+  visibleMeasures?: string[];
+  visibleDimensions?: string[];
+  leaderboardMeasureName?: string;
+
+  dashboardSortType?: SortType;
+  sortDirection?: SortDirection;
+};
+
+function persistentDashboardActions(
+  update: (this: void, updater: Updater<PersistentDashboardState>) => void,
+) {
+  function updateKey<K extends keyof PersistentDashboardState>(key: K) {
+    return (val: PersistentDashboardState[K]) => {
+      update((lup) => {
+        lup[key] = val;
+        return lup;
+      });
+    };
+  }
+
+  return {
+    updateVisibleMeasures: updateKey("visibleMeasures"),
+    updateVisibleDimensions: updateKey("visibleDimensions"),
+    updateLeaderboardMeasureName: updateKey("leaderboardMeasureName"),
+    updateDashboardSortType: updateKey("dashboardSortType"),
+    updateSortDirection: updateKey("sortDirection"),
+    reset() {
+      // cleanup dashboard settings. note that `timeZone` is not reset.
+      // it is intentional because it is an old feature
+      update((pd) => {
+        delete pd.visibleMeasures;
+        delete pd.visibleDimensions;
+        delete pd.leaderboardMeasureName;
+        delete pd.dashboardSortType;
+        delete pd.sortDirection;
+        return pd;
+      });
+    },
+  };
+}
+
+export type PersistentDashboardStore = Readable<PersistentDashboardState> &
+  ReturnType<typeof persistentDashboardActions>;
+export function createPersistentDashboardStore(storeKey: string) {
```

Comment:
Rather than creating a new store, did you consider taking our existing dashboard store and persisting it to local storage? Pros/cons? I'm a bit hesitant that we'll have to keep two stores in sync with the same data.

Reply:
A small section of the store is to be persisted. Check the details section for the exact fields. So it made sense to create a new store.

Reply:
Got it, I didn't realize that only a subset of the store is to be persisted.

Though I'm still wondering if it'd be simpler to persist the store in its entirety, then we can just choose which fields to restore in the `restorePersistedDashboardState()` function. Then it'd be easy to add/subtract fields we want to restore by just editing that one function.

Else, needing to remember to update both stores, I worry the two stores will often get out-of-sync and the feature will appear buggy.

Reply:
Another option to avoid synchronization issues – the persisted dashboard store could be a `derived` store that consumes the root dashboard store

Reply:
True. We can do that once the immer solution is in.

Reply:
Okay, just opened up an issue to track this: https://github.com/rilldata/rill-private-issues/issues/164